src/main/scala/nl/knaw/dans/lib/error/package.scala

```diff
@@ -76,4 +76,145 @@ package object error {
+        .getOrElse(Success(stream.map(_.get)))
+    }
+  }
+
+  implicit class TryExtensions[T](val t: Try[T]) extends AnyVal {
+    /**
+     * Applies the given side effecting function if and only if this is a `Success`.
+     *
+     * Example:
+     * {{{
+     *   import nl.knaw.dans.lib.error.TryExtensions
+     *
+     *   import scala.util.{Failure, Success, Try}
+     *
+     *   def getFileLength(file: File): Try[Long] =
+     *     if (file.exists) Success(file.length)
+     *     else Failure(new FileNotFoundException())
+     *
+     *   def performSideEffect(size: Long): Unit = println(s"size = $$size")
+     *
+     *   // Fill in existing or non-existing file
+     *   val file = new File("x")
+     *
+     *   getFileLength(file)
+     *     .doIfSuccess(size => performSideEffect(size))
+     * }}}
+     *
+     * @param f the side effecting function to be applied
+     * @return the original `Try`
+     */
+    def doIfSuccess(f: T => Unit): Try[T] = {
+      t match {
+        case success @ Success(x) => Try {
+          f(x)
+          return success
+        }
+        case e => e
+      }
+    }
+
+    /**
+     * Applies the given side effecting `PartialFunction` if and only if this is a `Failure` and
+     * the `Throwable` is defined in the `PartialFunction`.
+     *
+     * Example:
+     * {{{
+     *   import nl.knaw.dans.lib.error.TryExtensions
+     *
+     *   import scala.util.{Failure, Success, Try}
+     *
+     *   def getFileLength(file: File): Try[Long] =
+     *     if (file.exists) Success(file.length)
+     *     else Failure(new FileNotFoundException())
+     *
+     *   // Fill in existing or non-existing file
+     *   val file = new File("x")
+     *
+     *   getFileLength(file)
+     *     .doIfFailure {
+     *       case e: FileNotFoundException => println(e.getMessage)
+     *     }
+     * }}}
+     *
+     * @param f the side effecting function to be applied
+     * @return the original `Try`
+     */
+    def doIfFailure(f: PartialFunction[Throwable, Unit]): Try[T] = {
+      t match {
+        case failure @ Failure(e) if f.isDefinedAt(e) => Try {
+          f(e)
+          return failure
+        }
+        case x => x
+      }
+    }
+
+    /**
+     * Terminating operator for `Try` that converts the `Failure` case in a value.
+     *
+     * Example:
+     * {{{
+     *   import nl.knaw.dans.lib.error.TryExtensions
+     *
+     *   import scala.util.{Failure, Success, Try}
+     *
+     *   def getFileLength(file: File): Try[Long] =
+     *     if (file.exists) Success(file.length)
+     *     else Failure(new FileNotFoundException())
+     *
+     *   // Fill in existing or non-existing file
+     *   val file = new File("x")
+     *
+     *   getFileLength(file)
+     *     .getOrRecover {
+     *       // error codes
+     *       case _: FileNotFoundException => -1
+     *       case _ => -99
```

Comment:
I had a hard time coming up with a concise and simple to understand example. Maybe y'all can help me with this, so let me explain the purpose of this operator:

Let's say you're in the servlet part of a webservice (for example using Scalatra, as we do in our projects). You did a call to your service logic, which yielded a `Try[T]`. If this result is a `Success` you want to send an `Ok` with the `T` inside (so a response with status code 200 and content the instance of `T`. However, if the result is a `Failure`, we want to send some error code (maybe 404 with the message in the `Failure`). We can't use `recover` or `recoverWith` since this still yields a `Try`. Instead we want to break out of the `Try` context as we do in `getOrElse`. This method however does not allow us to act depending on the type of error. So, maybe we want to return a 404 for exception A, whereas we want to return 418 for exception B. That is exactly what this operator does: it uses the `getOrElse` idea of safely going out of the `Try` context and uses the `recover`/`recoverWith` idea to discriminate between the various kinds of exceptions.

You can see an example usage in https://github.com/DANS-KNAW/easy-bag-index/blob/master/src/main/scala/nl/knaw/dans/easy/bagindex/service/BagIndexServlet.scala#L46, where this operator is known as `onError` (to be replaced with this operator once the lib releases a new version).

Hope this makes the purpose a little clearer. Curious to so if we can come up with a better/clearer example than the one presented here.